stl/src/filesystem.cpp

```diff
@@ -172,6 +172,16 @@ namespace {
 
         return __std_win_error{GetLastError()};
     }
+
+    [[nodiscard]] bool __stdcall _Is_same_file(_In_ HANDLE _Handle_a, _In_ HANDLE _Handle_b) {
+        constexpr auto _Flags = __std_fs_file_flags::_Backup_semantics | __std_fs_file_flags::_Open_reparse_point;
+        FILE_ID_INFO Id_a, Id_b;
+        if (GetFileInformationByHandleEx(_Handle_a, FileIdInfo, &Id_a, sizeof(FILE_ID_INFO))

```

Comment:
Note that `FILE_ID_INFO` is not available on Windows 7. Consider falling back to `GetFileInformationByHandle` and examining `BY_HANDLE_FILE_INFORMATION::nFileIndexLow/High`.

Reply:
Yes, probably you can use:

`__std_fs_get_file_id` function: https://github.com/microsoft/STL/blob/bd7adb4a932725f60ba096580c415616486ab64c/stl/src/filesystem.cpp#L414-L454

and `struct __std_fs_file_id`: https://github.com/microsoft/STL/blob/ccc5aaaadc6dd1a1f80fb4d0755aff468817a274/stl/inc/xfilesystem_abi.h#L215-L218

Reply:
> Note that `FILE_ID_INFO` is not available on Windows 7. Consider falling back to `GetFileInformationByHandle` and examining `BY_HANDLE_FILE_INFORMATION::nFileIndexLow/High`.

I have a few questions regarding `BY_HANDLE_FILE_INFORMATION::nFileIndexLow/High` and `FILE_ID_INFO::FileId`.
According to the docs both of them represent some sort of **File Ids**. I did some testing with a bunch of random files across different volumes, I found that `BY_HANDLE_FILE_INFORMATION::nFileIndexLow/High` , `FILE_ID_INFO::FileId.Identifier[0..7]` are different and `FILE_ID_INFO::FileId.Identifier[8..15]` is always 0. `BY_HANDLE_FILE_INFORMATION::dwVolumeSerialNumber` and `FILE_ID_INFO::VolumeSerialNumber` are also different.
Can anyone explain me why is this?

For checking whether the files are hardlinks I think it is better to stick to `BY_HANDLE_FILE_INFORMATION::nFileIndexLow/High` for Windows 7 compatibilty.

Regarding the use of `__std_fs_get_file_id`, I am still awaiting a response to this https://github.com/microsoft/STL/pull/2062#issuecomment-880941498 .

Reply:
On NTFS file ids are 64-bit values. ReFS has 128-bit file ids. FILE_ID_INFO etc. were introduced to support the expanded file id space needed by ReFS, but on NTFS the ids remain 64-bit values and the upper 64-bits will always be zero. In order to properly support all edge cases on ReFS volumes, it is best to use the 128-bit file id type on systems that support it.

Similarly, the legacy structure has a 32-bit VSN field and the new structure returns the full 64-bit VSN - the different values are due to truncation.